File: services/tournament_service.py
import random
from database import db

class TournamentService:
    def __init__(self):
        self.current_tournament = {
            "registrations": [],
            "bracket": None,
            "status": 'open', # open, active, closed
            "winner": None
        }

    async def register_user(self, user_id, username):
        if self.current_tournament['status'] != 'open':
            return {"success": False, "msg": "Registration is closed for today."}
        
        if any(r['id'] == user_id for r in self.current_tournament['registrations']):
            return {"success": False, "msg": "You're already in the bracket!"}
        
        self.current_tournament['registrations'].append({"id": user_id, "username": username})
        return {"success": True, "msg": "✅ Succesfully enrolled in the Zenin Tournament! Bracket starts at 12:00 UTC."}

    async def generate_bracket(self):
        participants = self.current_tournament['registrations']
        if len(participants) < 2:
            self.current_tournament['status'] = 'closed'
            return {"success": False, "msg": "Not enough participants to start."}

        self.current_tournament['status'] = 'active'
        random.shuffle(participants)
        
        matches = []
        for i in range(0, len(participants), 2):
            if i + 1 < len(participants):
                matches.append({"p1": participants[i], "p2": participants[i+1], "winner": None})
            else:
                matches.append({"p1": participants[i], "p2": {"id": 0, "username": 'BYE'}, "winner": participants[i]})
        
        self.current_tournament['bracket'] = {
            "round": 1,
            "matches": matches
        }
        return {"success": True, "bracket": self.current_tournament['bracket']}

    def get_status(self):
        return {
            "count": len(self.current_tournament['registrations']),
            "status": self.current_tournament['status'],
            "bracket": self.current_tournament['bracket']
        }

tournament_service = TournamentService()

File: services/tournamentService.js
const db = require('../database');

/**
 * Tournament Service: Manages automated daily brackets.
 */
let currentTournament = {
    registrations: [],
    bracket: null,
    status: 'open', // open, active, closed
    winner: null
};

module.exports = {
    async registerUser(userId, username) {
        if (currentTournament.status !== 'open') return { success: false, msg: "Registration is closed for today." };
        if (currentTournament.registrations.includes(userId)) return { success: false, msg: "You're already in the bracket!" };
        
        currentTournament.registrations.push({ id: userId, username });
        return { success: true, msg: "✅ Succesfully enrolled in the Zenin Tournament! Bracket starts at 12:00 UTC." };
    },

    async generateBracket() {
        if (currentTournament.registrations.length < 2) {
            currentTournament.status = 'closed';
            return { success: false, msg: "Not enough participants to start." };
        }

        currentTournament.status = 'active';
        // Simple power-of-2 bracket seeding
        const participants = currentTournament.registrations.sort(() => 0.5 - Math.random());
        currentTournament.bracket = {
            round: 1,
            matches: []
        };

        for (let i = 0; i < participants.length; i += 2) {
            if (participants[i+1]) {
                currentTournament.bracket.matches.push({ p1: participants[i], p2: participants[i+1], winner: null });
            } else {
                currentTournament.bracket.matches.push({ p1: participants[i], p2: { id: 0, username: 'BYE' }, winner: participants[i] });
            }
        }
        return { success: true, bracket: currentTournament.bracket };
    },

    async resolveNextMatch() {
        if (currentTournament.status !== 'active') return null;
        
        const match = currentTournament.bracket.matches.find(m => !m.winner);
        if (!match) {
            // Round complete, generate next round? 
            // Simplified: Just crown a winner from the last matches for now
            return null; 
        }

        // Simulate resolution
        match.winner = Math.random() > 0.5 ? match.p1 : match.p2;
        return match;
    },

    getStatus() {
        return {
            count: currentTournament.registrations.length,
            status: currentTournament.status,
            bracket: currentTournament.bracket
        };
    }
};
